bfdd/bfdd_frr.c

```diff
@@ -0,0 +1,722 @@
+		} else if (strcmp(key, "echo-mode") == 0) {
+			bn->bn_bpc.bpc_echo = json_object_get_boolean(jo_val);
+		} else if (strcmp(key, "label") == 0) {
+			strlcpy(bn->bn_bpc.bpc_label,
+				json_object_get_string(jo_val),
+				sizeof(bn->bn_bpc.bpc_label));
+			bn->bn_bpc.bpc_has_label = true;
+		} else if (strcmp(key, "remote-detect-multiplier") == 0) {
+			bn->bn_bpc.bpc_remote_detectmultiplier =
+				json_object_get_int64(jo_val);
+		} else if (strcmp(key, "remote-receive-interval") == 0) {
+			bn->bn_bpc.bpc_remote_recvinterval =
+				json_object_get_int64(jo_val);
+		} else if (strcmp(key, "remote-transmit-interval") == 0) {
+			bn->bn_bpc.bpc_remote_txinterval =
+				json_object_get_int64(jo_val);
+		} else if (strcmp(key, "remote-echo-interval") == 0) {
+			bn->bn_bpc.bpc_remote_echointerval =
+				json_object_get_int64(jo_val);
+		}
+	}
+}
+
+
+/*
+ * Socket IO
+ */
+void bfdd_receive_debug(struct bfd_control_msg *bcm);
+
+void bfdd_receive_debug(struct bfd_control_msg *bcm)
+{
+	switch (bcm->bcm_type) {
+	case BMT_RESPONSE:
+		zlog_debug("%s: id: %d, type %s, length: %u, data: %s",
+			   __func__, ntohs(bcm->bcm_id), "RESPONSE",
+			   ntohl(bcm->bcm_length), bcm->bcm_data);
+		break;
+
+	case BMT_NOTIFY:
+		zlog_debug("%s: id: %d, type %s, length: %u, data: %s",
+			   __func__, ntohs(bcm->bcm_id), "NOTIFY",
+			   ntohl(bcm->bcm_length), bcm->bcm_data);
+		break;
+
+	case BMT_NOTIFY_ADD:
+	case BMT_NOTIFY_DEL:
+	case BMT_REQUEST_ADD:
+	case BMT_REQUEST_DEL:
+	default:
+		zlog_debug("%s: invalid response type (%d)\n", __func__,
+			   bcm->bcm_type);
+	}
+}
+
+/* Generic BFD function to handle notification messages. */
+int bfdd_receive_notification(struct bfd_control_msg *bcm, bool *repeat,
+			      void *arg)
+{
+	struct json_object *notification, *jo_val;
+	const char *sval;
+
+	/* Report unhandled versions */
+	switch (bcm->bcm_ver) {
+	case BMV_VERSION_1:
+		/* NOTHING */
+		break;
+
+	default:
+		zlog_debug("%s:%d: received unsupported version: %d", __func__,
+			   __LINE__, bcm->bcm_ver);
+		break;
+	}
+
+	/* This is not the response we are waiting. */
+	if (ntohs(bcm->bcm_id) != 0 || bcm->bcm_type != BMT_NOTIFY) {
+		zlog_debug("%s:%d: received non-notification packet", __func__,
+			   __LINE__);
+		bfdd_receive_debug(bcm);
+		return 0;
+	}
+
+	notification = json_tokener_parse((const char *)bcm->bcm_data);
+	if (json_object_object_get_ex(notification, "op", &jo_val) == false) {
+		zlog_debug("%s:%d: no operation described", __func__, __LINE__);
+		return 0;
+	}
+
+	sval = json_object_get_string(jo_val);
+	if (strcmp(sval, BCM_NOTIFY_PEER_STATUS) == 0) {
+		bfdd_peer_notification(notification);
+	} else if (strcmp(sval, BCM_NOTIFY_CONFIG_ADD) == 0
+		   || strcmp(sval, BCM_NOTIFY_CONFIG_DELETE) == 0
+		   || strcmp(sval, BCM_NOTIFY_CONFIG_UPDATE) == 0) {
+		bfdd_config_notification(notification);
+	}
+
+	return 0;
+}
+
+
+/*
+ * Auxiliary functions.
+ */
+void prefix2sa(const struct prefix *p, struct sockaddr_any *sa)
```

Comment:
This probably belongs in `lib/prefix.c`, once you're done syncing with the external daemon.